File: impls/tsbun/step3_env.ts
import { read_str, read_str_antlr } from "./reader.ts";
import { pr_str, pr_str_antlr } from "./printer.ts";
import {Env} from "./env.ts";

const eval_antlr = (ast: any, env: any): any => {
  return ast.eval(env);
}
const rootEnv = new Env();
rootEnv.set("+", (a: number, b: number) => a + b);
rootEnv.set("-", (a: number, b: number) => a - b);
rootEnv.set("*", (a: number, b: number) => a * b);
rootEnv.set("/", (a: number, b: number) => a / b);

const READ = (str: string): any => read_str_antlr(str);
const EVAL = (ast: any, _env?: any): any => eval_antlr(ast, rootEnv);
const PRINT = (exp: any): string => pr_str_antlr(exp)
const rep = (str: string): string => PRINT(EVAL(READ(str)));

const repl = async () => {
  const prompt = "user> ";
  process.stdout.write(prompt);
  for await (const line of console) {
    if (line.length === 0) {
      break;
    }
    try {
      process.stdout.write(rep(line));
      process.stdout.write("\n");
    } catch (e) {
      const err: Error = e;
      process.stdout.write(err.message);
      process.stdout.write("\n");
    }
    process.stdout.write(prompt);
  }
}
repl();

File: impls/elixir2/lib/mix/tasks/step3_env.ex

defmodule Mix.Tasks.Step3Env do

  def run(_) do
    env = Mal.Env.new()
    Mal.Env.set(env, "+", fn (a, b) -> a + b end)
    Mal.Env.set(env, "-", fn (a, b) -> a - b end)
    Mal.Env.set(env, "*", fn (a, b) -> a * b end)
    Mal.Env.set(env, "/", fn (a, b) -> a / b end)
    read_eval_print("(let* [x 10 y 20] (+ x y))", env)
    loop(env)
  end

  defp loop(env) do
    IO.write(:stdio, "user> ")
    IO.read(:stdio, :line)
      |> (&read_eval_print(&1, env)).()
      |> IO.puts

    loop(env)
  end

  defp read(input) do
    Mal.Reader.read_str(input)
  end

  defp eval_ast({:symbol, ast}, env) do
    case Mal.Env.get(env, ast) do
      {:ok, val} -> val
      :not_found -> throw({:error , "Symbol #{ast} not found"})
    end
  end
  defp eval_ast({:list, ast}, env) do
    {:list, Enum.map(ast, &eval(&1, env))}
  end
  defp eval_ast({:vector, ast}, env), do: {:vector, Enum.map(ast, &eval(&1, env))}
  defp eval_ast({:map, ast}, env) do
    map = for {key, value} <- ast, into: %{} do
      {key, eval(value, env)}
    end
    {:map, map}
  end
  defp eval_ast(ast, _env), do: ast

  defp eval({:list, []} = empty_list, _env), do: empty_list

  defp eval({:list, ast}, env), do: eval_list(ast, env)

  defp eval(input, env) do
    eval_ast(input, env)
  end

  # Builtin Funcs
  defp eval_list([{:symbol, "def!"}, {:symbol, key}, value], env) do
    evaled = eval(value, env)
    Mal.Env.set(env, key, evaled)
    evaled
  end


  defp eval_list([{:symbol, "let*"}, {_, bindings}, body], env) do
    new_env = Mal.Env.new(env)
    eval_bindings(bindings, new_env)
    eval(body, new_env)
  end

  defp eval_list(ast, env) do
    {:list, [func | args]} = eval_ast({:list, ast}, env)
    apply(func, args)
  end

  defp eval_bindings([], _), do: []
  defp eval_bindings([{:symbol, key}, value | rest], env) do
    Mal.Env.set(env, key, eval(value, env))
    eval_bindings(rest, env)
  end

  defp print(input) do
    Mal.Printer.print_str(input)
  end

  defp read_eval_print(:eof, _), do: exit(:normal)
  defp read_eval_print(line, env) do
    # env = %{
    #   "+" => fn (a, b) -> a + b end,
    #   "-" => fn (a, b) -> a - b end,
    #   "*" => fn (a, b) -> a * b end,
    #   "/" => fn (a, b) -> a / b end,
    # }

    read(line)
      |> (&eval(&1, env)).()
      |> print
  catch
    {:error, msg} -> "Error: #{msg}"
  end

end
